### PR TITLE
Implement chat message signing for 1.19.1+

### PR DESCRIPTION
**Pull request by mdcfe**
### Information

This PR refactors EssentialsX Chat in order to support chat previews and signed chat messages in 1.19.1. 

### Details

**Environments tested:**

- [x] Most recent Paper version (1.19.1, git-Paper-102)
- [x] Paper 1.18.2 (git-Paper-387)

**Demonstration:**  

Regular formatted chat is signed
Local chat, /toggleshout and shout/question prefix messages all signed

### CODE REVIEW

File: EssentialsChat/src/main/java/com/earth2me/essentials/chat/processing/AbstractChatHandler.java [hunk 1]
```diff
@@ -0,0 +1,269 @@
...
+    protected final ChatProcessingCache cache;
+
+    protected AbstractChatHandler(Essentials ess, EssentialsChat essChat) {
+        this.ess = ess;
+        this.essChat = essChat;
+        this.server = ess.getServer();
+        this.cache = new ChatProcessingCache();
+    }
+
+    // The initial chat formatting logic, handled at LOWEST priority
```
Comment (JRoy) [inline EssentialsChat/src/main/java/com/earth2me/essentials/chat/processing/AbstractChatHandler.java]:
```suggestion
    /*
    * The initial chat formatting logic, handled at LOWEST priority
    */
```

File: EssentialsChat/src/main/java/com/earth2me/essentials/chat/processing/AbstractChatHandler.java [hunk 2]
```diff
@@ -0,0 +1,269 @@
...
+        format = format.replace("{9}", nickname == null ? username : nickname);
+        synchronized (format) {
+            event.setFormat(format);
+        }
+
+        chat.setFormatResult(format);
+        chat.setMessageResult(event.getMessage());
+    }
+
+    // Local chat recipients logic, handled at NORMAL level
```
Comment (JRoy) [inline EssentialsChat/src/main/java/com/earth2me/essentials/chat/processing/AbstractChatHandler.java]:
```suggestion
    /*
    * Local chat recipients logic, handled at NORMAL level
    */
```

File: EssentialsChat/src/main/java/com/earth2me/essentials/chat/processing/AbstractChatHandler.java [hunk 3]
```diff
@@ -0,0 +1,269 @@
...
+        server.getPluginManager().callEvent(spyEvent);
+
+        if (!spyEvent.isCancelled()) {
+            for (final Player onlinePlayer : spyEvent.getRecipients()) {
+                onlinePlayer.sendMessage(String.format(spyEvent.getFormat(), user.getDisplayName(), spyEvent.getMessage()));
+            }
+        }
+    }
+
+    // Finalising the intermediate stages of chat processing, handled at HIGHEST level during previews
```
Comment (JRoy) [inline EssentialsChat/src/main/java/com/earth2me/essentials/chat/processing/AbstractChatHandler.java]:
```suggestion
    /*
    * Finalizing the intermediate stages of chat processing, handled at HIGHEST level during previews
    */
```

File: EssentialsChat/src/main/java/com/earth2me/essentials/chat/processing/AbstractChatHandler.java [hunk 4]
```diff
@@ -0,0 +1,269 @@
...
+        } else if (user.isToggleShout()) {
+            return message.length() > 1 ? "shout" : "";
+        } else {
+            return "";
+        }
+    }
+
+    private void charge(final User user, final Trade charge) throws ChargeException {
+        charge.charge(user);
+    }
```
Comment (JRoy) [inline EssentialsChat/src/main/java/com/earth2me/essentials/chat/processing/AbstractChatHandler.java:286]: what's the point of this method

File: EssentialsChat/src/main/java/com/earth2me/essentials/chat/processing/SignedChatHandler.java [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+        if (VersionUtil.getServerBukkitVersion().isLowerThan(VersionUtil.v1_19_1_R01)) {
+            return false;
+        }
+
+        try {
+            final Class<?> previewClass = Class.forName("org.bukkit.event.player.AsyncPlayerChatPreviewEvent");
+            if (!AsyncPlayerChatEvent.class.isAssignableFrom(previewClass)) {
+                essChat.getLogger().severe(I18n.tl("essChatNoSecureMsg", essChat.getDescription().getVersion()));
+                return false;
+            }
```
Comment (JRoy) [inline EssentialsChat/src/main/java/com/earth2me/essentials/chat/processing/SignedChatHandler.java:30]: Is this check needed? Why wouldn't this be the case?

**Review by JRoy — CHANGES_REQUESTED** [review]
first skim

File: EssentialsChat/src/main/java/com/earth2me/essentials/chat/processing/SignedChatHandler.java (hunk 5 above)
Comment (mdcfe) [reply EssentialsChat/src/main/java/com/earth2me/essentials/chat/processing/SignedChatHandler.java:30]: This is a draft API, and it's entirely possible Spigot might separate these events in the future (ie to provide the player's signature in the chat event, which wouldn't be relevant in the preview event). This just handles that possibility without blowing up.

File: EssentialsChat/src/main/java/com/earth2me/essentials/chat/processing/AbstractChatHandler.java (hunk 4 above)
Comment (mdcfe) [reply EssentialsChat/src/main/java/com/earth2me/essentials/chat/processing/AbstractChatHandler.java:286]: *legacy code*

File: EssentialsChat/src/main/java/com/earth2me/essentials/chat/processing/AbstractChatHandler.java [hunk 6]
```diff
@@ -0,0 +1,306 @@
...
+    protected final ChatProcessingCache cache;
+
+    protected AbstractChatHandler(Essentials ess, EssentialsChat essChat) {
+        this.ess = ess;
+        this.essChat = essChat;
+        this.server = ess.getServer();
+        this.cache = new ChatProcessingCache();
+    }
+
+    // Chat formatting logic (config and chat types), handled at LOWEST priority
```
Comment (mdcfe) [inline EssentialsChat/src/main/java/com/earth2me/essentials/chat/processing/AbstractChatHandler.java]: oops

**Review by JRoy — APPROVED** [review]

**Comment by mdcfe** [comment]
Known issue: this inadvertently reformats `/minecraft:tell`. There's not much we can do about this in 2.19.x.

**Comment by Pantera07** [comment]
```
[17:46:50] [Async Chat Thread - #0/INFO]: Mad_Daniel: test
[17:46:57] [Server thread/INFO]: [Essentials] CONSOLE issued server command: /ess version
[17:46:57] [Server thread/INFO]: Server version: 1.19.2-R0.1-SNAPSHOT git-Paper-113 (MC: 1.19.2)
[17:46:57] [Server thread/INFO]: Brand version: Paper
[17:46:57] [Server thread/INFO]: EssentialsX version: 2.19.5-dev+38-6dfa18c
[17:46:57] [Server thread/INFO]: Vault version: 1.7.3-b131
[17:46:57] [Server thread/INFO]: EssentialsXChat version: 2.19.5-dev+38-6dfa18c
[17:46:57] [Server thread/INFO]: Economy Layer: None
[17:46:57] [Server thread/INFO]: Fetching version information...
[17:46:57] [Craft Scheduler Thread - 2 - Essentials/INFO]: You're running the latest EssentialsX dev build!
[17:50:03] [Server thread/INFO]: Plugins (3): Essentials, EssentialsChat, Vault
```
How can I solve this?

**Comment by mdcfe** [comment]
> [...]
> 
> How can I solve this?

@Pantera07 Did you enable previews in your `server.properties`?

**Comment by Pantera07** [comment]
> > [...]
> > How can I solve this?
> 
> @Pantera07 Did you enable previews in your `server.properties`?

thank! I enable previews and this issue is resolved.

**Pull request merged**